addons/payment_onsite/models/pos_config.py

```diff
@@ -0,0 +1,25 @@
+from odoo import models, fields, api, _
+
+
+class PosConfigOnSite(models.Model):
+    _inherit = 'pos.config'
+    delivery_carrier_id = fields.Many2one('delivery.carrier', string="Shipping method",
+                                          readonly=True,
+                                          compute='_compute_delivery_carrier',
+                                          store=True)
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        pos_configs = super().create(vals_list)
+        product = self.env.ref('payment_onsite.onsite_delivery_product', raise_if_not_found=False)
+        if not product:
+            return pos_configs
+
+        carriers = self.env['delivery.carrier'].create([{
+            'name': _(f'{pos_config.name} - on site'),
+            'product_id': product.id,
+        } for pos_config in pos_configs if not pos_configs.delivery_carrier_id])
+
+        for pos_config, carrier in zip(pos_configs, carriers):
+            pos_config.delivery_carrier_id = carrier
```

Comment:
You are filtering in the first loop (`create`) but not the second one (`zip`) which may result in a bad config